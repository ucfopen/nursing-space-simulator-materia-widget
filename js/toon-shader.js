AFRAME.registerShader('toon-shader', {
  schema: {
    "uDirLightPos": {type: 'vec3', default: '100 50 5', is: 'uniform' },
    "uDirLightColor": { type: 'color', default: 'black', is: 'uniform' },

    "uMaterialColor": { type: 'color', default: 'white', is: 'uniform' },

    "uKd": { type: "number", default: 0.7, is: "uniform" },
    "uBorder": { type: "number", default: 0.4, is: "uniform" }

  },
  vertexShader: [
    "varying vec3 vNormal;",
    "varying vec3 vViewPosition;",

    "void main() {",

      "gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );",
      "vNormal = normalize( normalMatrix * normal );",
      "vec4 mvPosition = modelViewMatrix * vec4( position, 1.0 );",
      "vViewPosition = -mvPosition.xyz;",

    "}"

  ].join("\n"),

  fragmentShader: [   
      "uniform vec3 uMaterialColor;",

      "uniform vec3 uDirLightPos;",
      "uniform vec3 uDirLightColor;",

      "uniform float uKd;",
      "uniform float uBorder;",

      "varying vec3 vNormal;",
      "varying vec3 vViewPosition;",

      "void main() {",

        // compute direction to light
        "vec4 lDirection = viewMatrix * vec4( uDirLightPos, 0.0 );",
        "vec3 lVector = normalize( lDirection.xyz );",

        // diffuse: N * L. Normal must be normalized, since it's interpolated.
        "vec3 normal = normalize( vNormal );",
        //was: "float diffuse = max( dot( normal, lVector ), 0.0);",
        // solution
        "float diffuse = dot( normal, lVector );",
        "if ( diffuse > 0.6 ) { diffuse = 1.0; }",
        "else if ( diffuse > -0.2 ) { diffuse = 0.7; }",
        "else { diffuse = 0.3; }",

        "gl_FragColor = vec4( uKd * uMaterialColor * uDirLightColor * diffuse, 1.0 );",

      "}"

    ].join("\n")
});