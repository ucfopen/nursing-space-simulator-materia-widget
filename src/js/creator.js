document.addEventListener("DOMContentLoaded", function(event) {

	var qset = {
		items: [],
		options: {}
	};

	var widget = {
		engineName: "",
		title: ""
	};

	var materiaInterface = {};
	
	qset.options.gridLoader = {
		grid: "[[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",\"0\",{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}],[{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null},{\"id\":\"wall-1\",\"rotation\":0,\"stickers\":null}]]",
		columns: 30,
		content:
		"wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0 wall-1.0",
		rows: 12
	};

	materiaInterface.initNewWidget = function(w) {
		console.log("new");
	};

	materiaInterface.initExistingWidget = function(
		title,
		widget,
		qset,
		version,
		baseUrl
	) {
		console.log("Existing!");
		document.getElementById("title").value = title;
	};

	materiaInterface.onSaveClicked = function() {
		var title = document.getElementById("title").value;
		Materia.CreatorCore.save(title, qset, 1);
	};

	materiaInterface.onSaveComplete = function() {
		console.log("save complete!");
		return true;
	};

	return Materia.CreatorCore.start(materiaInterface);
});
