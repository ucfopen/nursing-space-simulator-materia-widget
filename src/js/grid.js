// Helper function for loadGrid that parses a string to an object
function _getAssetInfo(gridValue) {
	if (gridValue === "0") return gridValue;

	const assetInfo = gridValue.split(".");

	return {
		id: assetInfo[0],
		rotation: parseInt(assetInfo[1]),
		stickers: assetInfo.length == 3 ? assetInfo[2].split(",") : null
	};
}

// Sets walls that border the entire scene to face inward. Helps users add wall assets to those walls.
function _setInitialWallRotations(grid) {
	const height = grid.length;
	const width = grid[0].length;

	let rotation = 0;
	for (let i = 0; i < height; i++) {
		for (let j = 0; j < width; j++) {
			if (grid[i][j] == "0") continue;
			rotation = 0;

			if (i == 0) grid[i][j].rotation = 270;
			else if (i == height - 1) grid[i][j].rotation = 90;
			if (j == 0) grid[i][j].rotation = 0;
			else if (j == width - 1) grid[i][j].rotation = 180;
		}
	}

	return grid;
}

function _isInBounds(grid, row, col) {
	if (row < 0 || row >= grid.length) return false;
	if (col < 0 || col >= grid[0].length) return false;

	return true;
}

/**
 * takes a string and creates a grid data structure that can be manipulated with the methods below
 *
 * @param {string} gridString string representation of the grid to be loaded
 * @param {int} rows number of rows in the grid
 * @param {int} columns number of columns in the grid
 *
 * @return array representation of the grid string
 */
export function loadGrid(gridString, rows, columns) {
	if (gridString === null || rows === null || columns === null) {
		return null;
	}

	let tempGrid = gridString.split(" ");
	//return _setInitialWallRotations(
	return (
		[...Array(rows)].map(() => {
			return tempGrid.splice(0, columns).map(gridItem => {
				return _getAssetInfo(gridItem);
			});
		})
	);
}

/**
 * changes a cell rotation by 90 degrees
 *
 * @param {arary} grid grid to be updated
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 *
 * @return updated grid
 */
export function rotateCell(grid, x, z) {
	if (grid === null || x === null || z === null) {
		return null;
	}

	const col = x,
		row = z;

	if (grid[row][col] && grid[row][col] !== "0") {
		grid[row][col].rotation = (grid[row][col].rotation - 90 + 360) % 360;
	}

	// rotates the stickers, if available
	if (grid[row][col].stickers)
		grid[row][col].stickers.unshift(grid[row][col].stickers.pop())

	return grid;
}

/**
 * removes an item from a grid
 *
 * @param {array} grid grid to be updated
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 *
 * @return updated grid
 */
export function deleteItem(grid, x, z) {
	if (grid === null || x === null || z === null) {
		return null;
	}

	const col = x,
		row = z;

	grid[row][col] = "0";

	return grid;
}

/**
 * inserts an item into a grid
 *
 * @param {array} grid grid to be updated
 * @param {string} itemId id of the item to be placed
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 * @param {int} rotation rotation of the item being placed
 *
 * @return updated grid
 */
export function insertItem(grid, itemId, x, z, rotation = 180, stickers = null) {
	if (grid === null || x === null || z === null) {
		return null;
	}

	const col = x,
		row = z;

	grid[row][col] =
		itemId === null
			? (grid[row][col] = "0")
			: {
					id: itemId,
					rotation: rotation,
					stickers: stickers
				};

	// check adjacent spots for stickers to remove if necessary
	if (["wall-1", "door-1", "window"].includes(itemId))
	{
		for (let side = 0; side < 4; side++)
		{
			let adjItem = getItemId(grid, x + 2 - side, z);
			if (side % 2 == 0) {
				adjItem = getItemId(grid, x, z + side - 1);
			}
			if (["wall-1", "door-1", "window"].includes(adjItem))
			{
				if (side % 2 == 0)
					getStickers(grid, x, z + side - 1, true)
				else
					getStickers(grid, x + 2 - side, z, true);
			}
		}
	}

	return grid;
}

/**
 * checks to see if an item is placed in a grid cell
 *
 * @param {array} grid grid to be checked
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 *
 * @return boolean
 */
export function isCellAvailable(grid, x, z) {
	if (grid === null || x === null || z === null) {
		return false;
	}

	const col = x,
		row = z;

	if (!_isInBounds(grid, row, col)) return false;

	return grid[row][col] === "0";
}

/**
 * returns the id of the item placed in a grid cell
 *
 * @param {array} grid grid to be checked
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 *
 * @return id of item in cell, "0" if it is is out of bounds
 */
export function getItemId(grid, col, row) {
	if (grid === null || col === null || row === null) {
		return null;
	}

	if (!_isInBounds(grid, row, col)) return "0";

	return grid[row][col] === "0" ? "0" : grid[row][col].id;
}

/**
 * returns the id of the item placed in a grid cell
 *
 * @param {array} grid grid to be checked
 * @param {int} col column of gridcell to be checked
 * @param {int} row row of gridcell to be checked
 *
 * @return rotation of item in cell
 */
export function getCellRotation(grid, col, row) {
	if (grid === null || col === null || row === null) {
		return null;
	}

	return grid[row][col] === "0" ? 180 : grid[row][col].rotation;
}

/**
 * returns two lists of valid X and Z coordinates for a wall to be extended
 * from a given point
 *
 * @param {array} grid grid to be checked
 * @param {int} x the x-component of the wall to be extended
 * @param {int} z the z-component of the wall to be extended
 *
 * @return two lists for valid X and Z points
 */
export function findValidExtends(grid, x, z) {
	let validX = [x];
	let validZ = [z];
	// Valid direction array in order: up, right, bottom, left
	// In third-person view, Z is the horizontal axis, X is vertical
	let dir = [true, true, true, true];
	let level = 1;
	while (dir[0] || dir[1] || dir[2] || dir[3])
	{
		if (dir[0]) // up
		{
			if (isCellAvailable(grid, x + level, z))
				validX.push(x + level);
			else
				dir[0] = false;
		}
		if (dir[1]) // right
		{
			if (isCellAvailable(grid, x, z + level))
				validZ.push(z + level);
			else
				dir[1] = false;
		}
		if (dir[2]) // down
		{
			if (isCellAvailable(grid, x - level, z))
				validX.push(x - level);
			else
				dir[2] = false;
		}
		if (dir[3]) // left
		{
			if (isCellAvailable(grid, x, z - level))
				validZ.push(z - level);
			else
				dir[3] = false;
		}
		level++;
	}
	return [validX, validZ];
}

export function insertWalls(grid, startX, startZ, endX, endZ) {
	// If moving in the Z direction
	if (startX == endX)
	{
		let x = endX;
		let z = Math.min(startZ, endZ);
		let end = Math.max(startZ, endZ);
		while (z <= end)
		{
			if (z != startZ)
				grid = insertItem(grid, "wall-1", x, z, 0);
			z++;
		}
	}
	else
	{
		let z = endZ;
		let x = Math.min(startX, endX);
		let end = Math.max(startX, endX);
		while (x <= end)
		{
			if (x != startX)
				grid = insertItem(grid, "wall-1", x, z, 0);
			x++;
		}
	}
	return grid;
}

export function getStickers(grid, x, z, checkAdj) {
	if (!grid[z][x].stickers)
		grid[z][x].stickers = ["0", "0", "0", "0"];

	if (checkAdj)
	{
		// Check adjacent spots for a wall and remove the sticker if necessary
		for (let side = 0; side < 4; side++)
		{
			let adjItem = getItemId(grid, x + 2 - side, z);
			if (side % 2 == 0) {
				adjItem = getItemId(grid, x, z + side - 1);
			}

			// The "X" is a flag to show that a sticker cannot go on this side
			if (adjItem != "0" && ["wall-1", "door-1", "window"].includes(adjItem)) {
				grid[z][x].stickers[side] = "X";
			}
			else if (grid[z][x].stickers[side] == "X") {
				grid[z][x].stickers[side] = "0"
			}
		}
	}

	return grid[z][x].stickers;
}

export function setSticker(grid, x, z, side, sticker) {
	if (grid[z][x].stickers == null)
		grid[z][x].stickers = ["0", "0", "0", "0"];

	grid[z][x].stickers[side] = sticker;
	return grid;
}
