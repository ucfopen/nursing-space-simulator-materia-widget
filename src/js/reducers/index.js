import { combineReducers } from "redux";
import Position from "./positionReducer";
import Menu from "./menuReducer";
import Grid from "./gridReducer";
import Tour from "./tourReducer";

const rootReducer = combineReducers({
	grid: Grid,
	menu: Menu,
	position: Position,
	tour: Tour
});

export default rootReducer;
