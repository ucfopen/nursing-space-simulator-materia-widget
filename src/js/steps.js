const beginning = [
	{
		title: "Welcome to the Nursing Simulation Builder!",
		text:
			"Select <strong>Next</strong> to continue or <strong>Skip</strong> to close the tour.",
		selector: ".vr-scene",
		position: "bottom",
		style: {
			hole: { visibility: "hidden" }
		}
	},
	{
		title: "The Simulation Space",
		text:
			"This is the empty space where various objects can be placed. You can place additional walls and doors to create individual rooms.",
		selector: ".a-canvas",
		position: "bottom",
		style: {
			beacon: {
				offsetY: -450
			}
		}
	},
	{
		title: "Moving the Camera",
		text:
			"The camera's position can be adjusted using the buttons in this panel.",
		selector: "#camera-controls",
		position: "left",
		allowClicksThruHole: "true"
	},
	{
		title: "Taking a Screenshot",
		text: "Click/touch this button to take a screenshot of the room.",
		selector: "#screenshot",
		position: "left",
		allowClicksThruHole: "true"
	},
	{
		title: "Categories",
		text:
			"Here are the categories of objects and medical equipment to choose from.",
		selector: "#categories-list",
		position: "top",
		allowClicksThruHole: "true"
	},
	{
		title: "Picking an Object",
		text: "Click/touch an item to select it.",
		selector: "#asset-picker button:first-child",
		position: "right",
		allowClicksThruHole: "true",
		style: {
			footer: { display: "none " }
		}
	}
];

const clickInScene = [
	{
		title: "Placing Objects",
		text: "Click/touch inside the room to place the object.",
		selector: ".a-canvas",
		position: "bottom",
		allowClicksThruHole: "true",
		style: {
			footer: { display: "none " }
		}
	}
];

const objectOptions = [
	{
		title: "Object Options",
		text:
			"After you place an object, you have the option to confirm the placement, rotate, or remove the object.",
		selector: "#UI-selected-asset-options",
		position: "right",
		allowClicksThruHole: "false"
	},
	{
		title: "Object Movement",
		text:
			"When an object is selected, the arrow keys can be used to move the object around inside the room. You can also click on another empty spot in the room to move it.",
		selector: "#camera-controls",
		position: "left",
		allowClicksThruHole: "true"
	},
	{
		title: "First-Person Viewer",
		text:
			"You can view the room from a more personal point of view by selecting the <strong>First-Person Viewer</strong>.",
		selector: "#vr-viewer-mode",
		position: "top",
		allowClicksThruHole: "true",
		style: {
			footer: { display: "none " }
		}
	}
];

const clickFirstPersonViewer = [
	{
		title: "Go Into First-Person View",
		text: "Click/touch inside the room to go into first-person mode.",
		selector: ".a-canvas",
		position: "bottom",
		allowClicksThruHole: "true",
		style: {
			footer: { display: "none " }
		}
	}
];

const clickCameraInScene = [
	{
		title: "Welcome to First Person",
		text:
			"Here you can look around the room you created by dragging the screen.",
		selector: ".vr-scene",
		position: "bottom",
		allowClicksThruHole: "true"
	},
	{
		title: "Going Back",
		text: "You can click/touch here to return to the previous view.",
		selector: "#back",
		position: "bottom",
		allowClicksThruHole: "true"
	},
	{
		title: "VR Mode",
		text:
			"If on a supported mobile device, You can touch here to enter VR Mode! <img id='vr-gif' align='middle' height='250' width='350' src='assets/cardboard_vr.gif'></img>",
		selector: ".a-enter-vr-button",
		position: "bottom",
		allowClicksThruHole: "true"
	}
];

const steps = {
	0: beginning,
	1: clickInScene,
	2: objectOptions,
	3: clickFirstPersonViewer,
	4: clickCameraInScene
};
export default steps;
