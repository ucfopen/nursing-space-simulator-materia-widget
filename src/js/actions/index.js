import { loadGrid } from "../grid";

export const INIT_DATA = "init_data";

export function initData(qset) {
	const grid = loadGrid(
		qset.options.gridLoader.content,
		qset.options.gridLoader.rows,
		qset.options.gridLoader.columns
	);

	const categories = qset.options.categories;
	const stickerTypes = qset.options.stickerTypes;
	const assets = qset.options.assets;

	return {
		type: INIT_DATA,
		payload: { grid, categories, stickerTypes, assets }
	};
}
