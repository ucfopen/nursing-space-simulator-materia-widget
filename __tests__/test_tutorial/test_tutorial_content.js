import * as steps from "../../src/js/steps";

describe("Tutorial Tests", () => {
	it("should have the correct part 1", () => {
		const part1JSON = JSON.stringify(steps.part1);
		expect(part1JSON).toMatchSnapshot();
	});

	it("should have the correct part 2", () => {
		const part2JSON = JSON.stringify(steps.part2);
		expect(part2JSON).toMatchSnapshot();
	});

	it("should have the correct part 3", () => {
		const part3JSON = JSON.stringify(steps.part3);
		expect(part3JSON).toMatchSnapshot();
	});

	it("should have the correct part 4", () => {
		const part4JSON = JSON.stringify(steps.part4);
		expect(part4JSON).toMatchSnapshot();
	});

	it("should have the correct part 5", () => {
		const part5JSON = JSON.stringify(steps.part5);
		expect(part5JSON).toMatchSnapshot();
	});
});
