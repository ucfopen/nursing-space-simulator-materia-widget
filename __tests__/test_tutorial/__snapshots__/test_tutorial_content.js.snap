// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tutorial Tests should have the correct part 1 1`] = `"[{\\"title\\":\\"Welcome to the Nursing Simulation Builder!\\",\\"text\\":\\"Select <strong>Next</strong> to continue or <strong>Skip</strong> to close the tour.\\",\\"selector\\":\\"#app\\",\\"position\\":\\"top\\",\\"style\\":{\\"hole\\":{\\"visibility\\":\\"hidden\\"},\\"arrow\\":{\\"display\\":\\"none\\"}}},{\\"title\\":\\"The Simulation Space\\",\\"text\\":\\"Below is the scene where various objects can be placed. You can place additional walls and doors to create rooms.\\",\\"selector\\":\\".vr-scene\\",\\"position\\":\\"top\\",\\"style\\":{\\"hole\\":{\\"visibility\\":\\"hidden\\"},\\"arrow\\":{\\"display\\":\\"none\\"}}},{\\"title\\":\\"Moving the Camera\\",\\"text\\":\\"The camera's position can be adjusted using these buttons. The middle button recenters the camera.\\",\\"selector\\":\\"#camera-controls\\",\\"position\\":\\"left\\",\\"allowClicksThruHole\\":true},{\\"title\\":\\"Taking a Screenshot\\",\\"text\\":\\"This button can be used to take a screenshot of the room.\\",\\"selector\\":\\"#screenshot\\",\\"position\\":\\"left\\",\\"allowClicksThruHole\\":true},{\\"title\\":\\"Categories\\",\\"text\\":\\"Here are the categories of objects to choose from. Make sure the <strong>Equipment</strong> category is selected.\\",\\"selector\\":\\"#categories-list\\",\\"position\\":\\"top\\",\\"allowClicksThruHole\\":true},{\\"title\\":\\"Picking an Object\\",\\"text\\":\\"<strong>Click/touch</strong> a bed now to select it.\\",\\"selector\\":\\"#asset-picker button:first-child\\",\\"position\\":\\"right\\",\\"allowClicksThruHole\\":true,\\"style\\":{\\"button\\":{\\"visibility\\":\\"hidden\\"}}}]"`;

exports[`Tutorial Tests should have the correct part 2 1`] = `"[{\\"title\\":\\"Placing an Object\\",\\"text\\":\\"<strong>Click/touch</strong> inside the room to place the bed.\\",\\"selector\\":\\".a-canvas\\",\\"position\\":\\"bottom\\",\\"allowClicksThruHole\\":true,\\"style\\":{\\"button\\":{\\"visibility\\":\\"hidden\\"}}}]"`;

exports[`Tutorial Tests should have the correct part 3 1`] = `"[{\\"title\\":\\"Object Options\\",\\"text\\":\\"After an object is placed, this option panel appears. Here you confirm placement, rotate, or remove the selected object.\\",\\"selector\\":\\"#UI-selected-asset-options\\",\\"position\\":\\"right\\",\\"allowClicksThruHole\\":true},{\\"title\\":\\"Object Movement\\",\\"text\\":\\"The arrow buttons can be used to move a selected object within the scene. Clicking/touching an empty space will also move the object. Additionally, if using the Chrome Browser, objects can be dragged around the scene.\\",\\"selector\\":\\".a-canvas\\",\\"position\\":\\"left\\",\\"allowClicksThruHole\\":true},{\\"title\\":\\"Placing Walls\\",\\"text\\":\\"Select the <strong>Construction</strong> category.\\",\\"selector\\":\\".category-construction\\",\\"position\\":\\"top\\",\\"allowClicksThruHole\\":true,\\"style\\":{\\"button\\":{\\"visibility\\":\\"hidden\\"}}}]"`;

exports[`Tutorial Tests should have the correct part 4 1`] = `"[{\\"title\\":\\"Placing Walls\\",\\"text\\":\\"<strong>Click/touch</strong> the wall to select it.\\",\\"selector\\":\\".asset-wall-1\\",\\"position\\":\\"right\\",\\"allowClicksThruHole\\":true,\\"style\\":{\\"button\\":{\\"visibility\\":\\"hidden\\"}}}]"`;

exports[`Tutorial Tests should have the correct part 5 1`] = `"[{\\"title\\":\\"Placing Walls\\",\\"text\\":\\"Now <strong>click/touch</strong> in the scene to place the wall.\\",\\"selector\\":\\".a-canvas\\",\\"position\\":\\"right\\",\\"allowClicksThruHole\\":true,\\"style\\":{\\"button\\":{\\"visibility\\":\\"hidden\\"}}}]"`;
